heron/metricscachemgr/src/java/com/twitter/heron/metricscachemgr/MetricsCacheManager.java

```diff
@@ -126,6 +129,61 @@ public MetricsCacheManager(String topologyName,
 
     // Construct the server to respond to query request
     metricsCacheManagerHttpServer = new MetricsCacheManagerHttpServer(metricsCache, statsPort);
+
+    this.mainThreadId = Thread.currentThread().getId();
+    // Add exception handler for any uncaught exception here.
+    Thread.setDefaultUncaughtExceptionHandler(new DefaultExceptionHandler());
+  }
+
+  /**
+   * Handler for catching exceptions thrown by any threads (owned either by topology or heron
+   * infrastructure). When one IMetricsSink throws uncaught exceptions, we would try to restart this
+   * sink unless we have hit the # of retry attempts. When Metrics Manager internal exceptions are
+   * caught or we have restart IMetricsSink with too many attempts, Metrics Manager would flush any
+   * remain logs and exit.
+   */
+  public class DefaultExceptionHandler implements Thread.UncaughtExceptionHandler {
+
+    /**
+     * Handler for uncaughtException
+     */
+    public void uncaughtException(Thread thread, Throwable exception) {
+      // Add try and catch block to prevent new exceptions stop the handling thread
+      try {
+        // Delegate to the actual one
+        handleException(thread, exception);
```

Comment:
I don't think we need to make it so complicated. We will not do any recovering here,
so we can simply:
```
Log(...);
// The ReleaseResourcesExplicitlyIfNeeded() can be a bit tricky. 
// If you want to release resources explicitly, make sure set a timeout to void unexpected hanging
// Also, it is possible that it can have other unexpected behaviors
// If all resources can eventually be released by the death of a process, 
// we are ok not to clean it explicitly
ReleaseResourcesExplicitlyIfNeeded(); 
Runtime.getRuntime().halt(1);
```